api/types/tunnel.go

```diff
@@ -123,19 +120,24 @@ func (r *ReverseTunnelV2) SetName(e string) {
 	r.Metadata.Name = e
 }
 
+// setStaticFields sets static resource header and metadata fields.
+func (r *ReverseTunnelV2) setStaticFields() {
+	r.Kind = KindReverseTunnel
+	r.Version = V2
+}
+
 // CheckAndSetDefaults checks and sets defaults
 func (r *ReverseTunnelV2) CheckAndSetDefaults() error {
-	err := r.Metadata.CheckAndSetDefaults()
-	if err != nil {
+	r.setStaticFields()
+	if err := r.Metadata.CheckAndSetDefaults(); err != nil {
 		return trace.Wrap(err)
 	}
-	if r.Version == "" {
-		r.Version = V2
-	}
 
-	err = r.Check()
-	if err != nil {
-		return trace.Wrap(err)
+	if strings.TrimSpace(r.Spec.ClusterName) == "" {
+		return trace.BadParameter("Reverse tunnel validation error: empty cluster name")
+	}
+	if len(r.Spec.DialAddrs) == 0 {
+		return trace.BadParameter("Invalid dial address for reverse tunnel '%v'", r.Spec.ClusterName)
```

Comment:
nit: https://github.com/golang/go/wiki/CodeReviewComments#error-strings
```suggestion
		return trace.BadParameter("reverse tunnel validation error: empty cluster name")
	}
	if len(r.Spec.DialAddrs) == 0 {
		return trace.BadParameter("invalid dial address for reverse tunnel %q", r.Spec.ClusterName)
```